src/kcas.mli

```diff
@@ -323,3 +349,85 @@ module Tx : sig
       transaction mechanism has no way to intelligently wait until shared memory
       locations are modified by other domains. *)
 end
+
+(** Explicit transaction log passing on shared memory locations. *)
+module Xt : sig
+  type 'x t
+  (** Type of an explicit transaction log on shared memory locations.
+
+      Note that a transaction log itself is not domain-safe and should generally
+      only be used by a single domain.  If a new domain is spawned inside a
+      function recording shared memory accesses to a log and the new domain also
+      records accesses to the log it may become inconsistent. *)
+
+  (** {1 Recording accesses}
+
+      Accesses of shared memory locations using an explicit transaction log
+      first ensure that the initial value of the shared memory location is
+      recorded in the log and then act on the current value of the shared memory
+      location as recorded in the log. *)
+
+  val get : xt:'x t -> 'a Loc.t -> 'a
+  (** [get ~xt r] returns the current value of the shared memory location [r] in
+      the explicit transaction log [xt]. *)
+
+  val set : xt:'x t -> 'a Loc.t -> 'a -> unit
+  (** [set ~xt r v] records the current value of the shared memory location [r]
+      to be the given value [v] in the explicit transaction log [xt]. *)
+
+  val update : xt:'x t -> 'a Loc.t -> ('a -> 'a) -> 'a
+  (** [update ~xt r f] is equivalent to [let x = get ~xt r in set ~xt r (f x); x]. *)
+
+  val modify : xt:'x t -> 'a Loc.t -> ('a -> 'a) -> unit
+  (** [modify ~xt r f] is equivalent to [let x = get ~xt r in set ~xt r (f x)]. *)
+
+  val exchange : xt:'x t -> 'a Loc.t -> 'a -> 'a
+  (** [exchange ~xt r v] is equivalent to [update ~xt r (fun _ -> v)]. *)
+
+  val fetch_and_add : xt:'c t -> int Loc.t -> int -> int
+  (** [fetch_and_add ~xt r n] is equivalent to [update ~xt r ((+) n)]. *)
+
+  val incr : xt:'x t -> int Loc.t -> unit
+  (** [incr ~xt r] is equivalent to [fetch_and_add ~xt r 1 |> ignore]. *)
+
+  val decr : xt:'x t -> int Loc.t -> unit
+  (** [decr ~xt r] is equivalent to [fetch_and_add ~xt r (-1) |> ignore]. *)
+
+  (** {1 Performing accesses} *)
+
+  type 'a tx = { tx : 'x. xt:'x t -> 'a }
```

Comment:
What's the advantage of using a label here? What's wrong with:

```ocaml
type 'a tx = { tx : 'x. x t -> 'a }
```

The exposed API (a record) is already a bit heavy - maybe there's value in not adding that extra label as well? There's perhaps a clever way to hide this behind a constructor. Another thing to consider here: exposing concrete types often leads to headaches for future API evolutions.

Reply:
The label helps to make the explicit log passing stand out (which I consider a good thing) and perhaps make it more idiomatic.  On the other hand, it also allows the label to be omitted (regardless of its position) as in a call like `Xt.commit { tx = Queue.pop_front_opt queue }`.

The reason for having the record is to require the function be polymorphic with respect to the log.  IOW, it is using rank-2 polymorphism.  This makes it difficult to accidentally leak or reuse the log (the `~xt` parameter), which would be unsafe.  IOW, an attempt to e.g. store the value `~xt` to a ref cell allocated outside the transaction and then reuse it in another transaction will be prevented by the type system.

Reply:
Ok, in that case, I suggest that the examples could use the idiomatic API.

i.e. in the example, always define an external function with an extra non-labelled parameter and use it. This way, it's extra-clear why you ask users to write such convoluted code :-)